### PR TITLE
[KEP-3280] Part1: Add New API DisruptionPolicy in Pod and PriorityClass

### PR DESCRIPTION
**Pull request by denkensk**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This is the first part of implementing KEP 3280, including api update and auto code generate.

#### Which issue(s) this PR fixes:

https://github.com/kubernetes/enhancements/issues/3280

#### Does this PR introduce a user-facing change?

```release-note
New API field `DisruptionPolicy` in Pod and PriorityClass  is introduced to enable users to control the behavior of the scheduler in considering the PDB during the preemption process.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md
```

### CODE REVIEW

**Comment by denkensk** [comment]
There are 2 PRs upcoming: scheduler implementation and integration/e2e tests.

**Comment by denkensk** [comment]
cc @kubernetes/sig-scheduling-leads

**Comment by alculquicondor** [comment]
You are posting this 3 business days before the code freeze. I'll do my best to review, but I can't promise much.

**Comment by denkensk** [comment]
> You are posting this 3 business days before the code freeze. I'll do my best to review, but I can't promise much.

Thanks. Aldo. Sorry for not submitting earlier, but at least hope to look good to everyone at the api definition.

File: pkg/api/pod/util.go [hunk 1]
```diff
@@ -682,6 +683,27 @@ func dropDisabledMatchLabelKeysField(podSpec, oldPodSpec *api.PodSpec) {
...
+		podSpec.DisruptionPolicy = nil
+	}
+}
+
+// podPriorityInUse returns true if the pod spec is non-nil and has Priority or PriorityClassName set.
+func podPriorityInUse(podSpec *api.PodSpec) bool {
+	if podSpec == nil {
+		return false
+	}
+	if podSpec.Priority != nil || podSpec.PriorityClassName != "" {
```
Comment (alculquicondor) [inline pkg/api/pod/util.go:701]: This should be checking the `DisruptionPolicy` field instead.

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -2175,6 +2175,41 @@ const (
 	PreemptNever PreemptionPolicy = "Never"
 )
 
+// DisruptionPolicy specifies how pods with this priority class may be disrupted when being preempted.
+type DisruptionPolicy struct {
```
Comment (alculquicondor) [inline pkg/apis/core/types.go:2179]: This is significantly different from what was agreed on in the KEP

https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens#design-details

I'm ok having a struct, but just with one field PriorityGreaterThanOrEqual

Comment (denkensk) [reply pkg/apis/core/types.go:2179]: When implementing this feature, I found that just one field PriorityGreaterThanOrEqual can't cover all situations. There are three situations listed below:

```go
const (
	// PDBIgnore indicates that the associated PDB will not be considered
	// when this Pod is preempted during scheduling.
	PDBIgnore PDBDisruptionPolicy = "PDBIgnore"
	// PDBBestEffort indicates that the associated PDB should be protected as
	// much as possible from being violated when this Pod is preempted during scheduling.
	PDBBestEffort PDBDisruptionPolicy = "PDBBestEffort"
	// PDBStrict indicates that the associated PDB should be protected strictly from
	// being violated when this Pod is preempted during scheduling.
	// But if the preempter has a priority class name of `system-cluster-critical` or
	// `system-node-critical`, may still potentially violate the victim's PDB.
	PDBStrict PDBDisruptionPolicy = "PDBStrict"
)
```

We may also need to provide an option for the user that the scheduler ignores the PDB when choosing the victims during scheduling. This allows users to make their choices clear and predictable. WDYT?

File: pkg/api/pod/util.go (hunk 1 above)
Comment (denkensk) [reply pkg/api/pod/util.go]: Updated

File: pkg/apis/core/types.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/apis/core/types.go:2179]: 1. A behavior like `PDBIgnore` was not part of the goals and I don't see why it's useful.
2. `PDBStrict` is basically the same as `PDBBestEffort` with `priorityGreaterThanOrEqual=value(system-node-critical)`

So I don't think it's useful.

**Pull request closed without merging**